File: examples/scripts/run_erc20.py
import json
import os
import sys
from eth_account import Account
from solcx import compile_standard, install_solc, get_installed_solc_versions
from web3 import Web3, HTTPProvider
from web3.middleware import geth_poa_middleware
sys.path.append(os.path.abspath(os.path.join(os.getcwd(), '../../tools/python')))
from crypto import decrypt, prepare_IT
import subprocess

SOLC_VERSION = '0.8.19'
FILE_PATH = 'ERC20TokenStandardContract.sol'
PROVIDER_URL = 'http://localhost:7000'
INITIAL_BALANCE = 500000000

def check_and_install_solc_version(solc_version):
    if solc_version not in get_installed_solc_versions():
        install_solc(solc_version)

def read_solidity_code(file_path):
    if not os.path.exists(file_path):
        raise Exception(f"The file {file_path} does not exist")
    with open(file_path, 'r') as file:
        return file.read()

def compile_solidity(solc_version, file_path, solidity_code, mpc_inst_path, mpc_core_path):
    compiled_sol = compile_standard({
        "language": "Solidity",
        "sources": {
            "MPCInst.sol": {"urls": [mpc_inst_path]},
            "MpcCore.sol": {"urls": [mpc_core_path]},
            file_path: {"content": solidity_code},
        },
        "settings": {
            "outputSelection": {
                "*": {"*": ["metadata", "evm.bytecode", "evm.bytecode.sourceMap"]}
            }
        },
    },
    solc_version=solc_version,
    allow_paths=[".", "../../lib/solidity"]
    )
    return compiled_sol

def deploy_contract(w3, account, contract_abi, contract_bytecode, gas_limit=10000000, gas_price='30'):
    contract = w3.eth.contract(abi=contract_abi, bytecode=contract_bytecode)
    construct_txn = contract.constructor("Soda", "SOD", INITIAL_BALANCE).build_transaction({
        'from': account.address,
        'chainId': 50505050,
        'nonce': w3.eth.get_transaction_count(account.address),
        'gas': gas_limit,
        'gasPrice': w3.to_wei(gas_price, 'gwei')
    })
    signed_txn = w3.eth.account.sign_transaction(construct_txn, account._private_key)
    tx_hash = w3.eth.send_raw_transaction(signed_txn.rawTransaction)
    tx_receipt = w3.eth.wait_for_transaction_receipt(tx_hash)
    print(f'Contract Deployed At Address: {tx_receipt.contractAddress}')
    return w3.eth.contract(address=tx_receipt.contractAddress, abi=contract_abi)


def execute_transaction(w3, account, function, gas_limit=2000000, gas_price='50'):
    transaction = function.build_transaction({
        'from': account.address,
        'chainId': 50505050,
        'gas': gas_limit,
        'gasPrice': w3.to_wei(gas_price, 'gwei'),
        'nonce': w3.eth.get_transaction_count(account.address),
    })
    signed_txn = w3.eth.account.sign_transaction(transaction, private_key=account._private_key)
    txn_hash = w3.eth.send_raw_transaction(signed_txn.rawTransaction)
    txn_receipt = w3.eth.wait_for_transaction_receipt(txn_hash)
    print(f'Transaction successful with hash: {txn_hash.hex()}')
    return txn_receipt


def check_expected_result(name, expected_result, result):
    if result == expected_result:
        print(f'Test {name} succeeded: {result}')
    else:
        raise ValueError(f'Test {name} failed. Expected: {expected_result}, Actual: {result}')


def decrypt_value(my_CTBalance, user_key):
    
    # Convert ct to bytes (big-endian)
    byte_array = my_CTBalance.to_bytes(32, byteorder='big')

    # Split ct into two 128-bit arrays r and cipher
    cipher = byte_array[:16]
    r = byte_array[16:]

    # Decrypt the cipher
    decrypted_message = decrypt(user_key, r, cipher)

    # Print the decrypted cipher
    decrypted_balance = int.from_bytes(decrypted_message, 'big')

    return decrypted_balance
    

def get_function_signature(function_abi):
    # Extract the input types from the ABI
    input_types = ','.join([param['type'] for param in function_abi.get('inputs', [])])

    # Generate the function signature
    return f"{function_abi['name']}({input_types})"

def execute_and_check_balance(w3, account, contract, function, user_key, name, expected_balance):
    execute_transaction(w3, account, function)
    # Get my encrypted balance, decrypt it and check if it matches the expected value
    my_CTBalance = contract.functions.balanceOf().call({'from': account.address})
    my_balance = decrypt_value(my_CTBalance, user_key)
    check_expected_result(name, expected_balance, my_balance)

def check_allowance(account, contract, user_key, expected_allowance):
    # Get my encrypted allowance, decrypt it and check if it matches the expected value
    allowanceCT = contract.functions.allowance(account.address, account.address).call({'from': account.address})
    allowance = decrypt_value(allowanceCT, user_key)
    check_expected_result('allowance', expected_allowance, allowance)

def check_expected_result(name, expected_result, result):
    if result == expected_result:
        print(f'Test {name} succeeded: {result}')
    else:
        raise ValueError(f'Test {name} failed. Expected: {expected_result}, Actual: {result}')

def main():
    check_and_install_solc_version(SOLC_VERSION)
    solidity_code = read_solidity_code("../contracts/" + FILE_PATH)

    mpc_inst_path = "../../lib/solidity/MPCInst.sol"
    mpc_core_path = "../../lib/solidity/MpcCore.sol"
    compiled_sol = compile_solidity(SOLC_VERSION, FILE_PATH, solidity_code, mpc_inst_path, mpc_core_path)

    contract_name = FILE_PATH.split('.')[0]
    bytecode = compiled_sol['contracts'][FILE_PATH][contract_name]['evm']['bytecode']['object']
    contract_abi = json.loads(compiled_sol['contracts'][FILE_PATH][contract_name]['metadata'])['output']['abi']
    contract_bytecode = f'0x{bytecode}'

    w3 = Web3(HTTPProvider(PROVIDER_URL))
    if not w3.is_connected():
        print("Failed to connect to the Ethereum node.")
        exit()

    w3.middleware_onion.inject(geth_poa_middleware, layer=0)

    private_key = os.environ.get('SIGNING_KEY')
    account = Account.from_key(private_key)
    w3.eth.default_account = account.address

    contract = deploy_contract(w3, account, contract_abi, contract_bytecode)

    print("************* View functions *************")
    name = contract.functions.name().call({'from': account.address})
    print("Function call result name:", name)

    symbol = contract.functions.symbol().call({'from': account.address})
    print("Function call result symbol:", symbol)

    decimals = contract.functions.decimals().call({'from': account.address})
    print("Function call result decimals:", decimals)

    totalSupply = contract.functions.totalSupply().call({'from': account.address})
    print("Function call result totalSupply:", totalSupply)

    user_key_hex = os.environ.get('USER_KEY')
    user_key = bytes.fromhex(user_key_hex)  


    print("************* Initial balance = ", INITIAL_BALANCE, " *************")
    my_CTBalance = contract.functions.balanceOf().call({'from': account.address})
    my_balance = decrypt_value(my_CTBalance, user_key)
    check_expected_result('balanceOf', INITIAL_BALANCE, my_balance)

    # Generate a new Ethereum account for Alice
    alice_address = Account.create()

    plaintext_integer = 5

    print("************* Transfer clear ", plaintext_integer, " to Alice *************")
    # Transfer 5 SOD to Alice
    function = contract.functions.transfer(alice_address.address, plaintext_integer, True)
    execute_and_check_balance(w3, account, contract, function, user_key, 'transfer', INITIAL_BALANCE - plaintext_integer)

    print("************* Transfer clear ", plaintext_integer, " to Alice *************")
    # Transfer 5 SOD to Alice
    function = contract.functions.contractTransferClear(alice_address.address, plaintext_integer)
    execute_and_check_balance(w3, account, contract, function, user_key, 'contract transfer clear', INITIAL_BALANCE - 2*plaintext_integer)

    print("************* Transfer IT ", plaintext_integer, " to Alice *************")
    # In order to generate the input text, we need to use some data of the function. 
    # For example, the address of the user, the address of the contract and also the function signature.
    # In order to get the function signature as simple as possible, we decided to use a dummy function with dummy inputs and use it to generate the signature.
    # After the signature is generated, we call prepare input text function and get the input text to use in the real function.
    dummyCT = 0
    dummySignature = bytes(65)
    function = contract.functions.transfer(alice_address.address, dummyCT, dummySignature, False)
    func_sig = get_function_signature(function.abi) # Get the function signature
    # Prepare the input text for the function
    ct, signature = prepare_IT(plaintext_integer, user_key, account, contract, func_sig, bytes.fromhex(private_key[2:]))
    # Create the real function using the prepared IT
    function = contract.functions.transfer(alice_address.address, ct, signature, False)
    # Transfer 5 SOD to Alice
    execute_and_check_balance(w3, account, contract, function, user_key, 'transfer IT', INITIAL_BALANCE - 3*plaintext_integer)

    print("************* Transfer clear ", plaintext_integer, " from my account to Alice without allowance *************")
    # Trying to transfer 5 SOD to Alice. There is no allowance, transfer should fail
    function = contract.functions.transferFrom(account.address, alice_address.address, plaintext_integer, True)
    execute_and_check_balance(w3, account, contract, function, user_key, 'transfer from', INITIAL_BALANCE - 3*plaintext_integer)

    print("************* Approve ", plaintext_integer*10, " to my address *************")
    # Set allowance for this account
    function = contract.functions.approveClear(account.address, plaintext_integer*10)
    execute_transaction(w3, account, function)
    # Get my allowance to check that the allowance is correct
    check_allowance(account, contract, user_key, plaintext_integer*10)

    print("************* Transfer clear ", plaintext_integer, " from my account to Alice *************")
    # Transfer 5 SOD to Alice
    function = contract.functions.transferFrom(account.address, alice_address.address, plaintext_integer, True)
    execute_and_check_balance(w3, account, contract, function, user_key, 'transfer from', INITIAL_BALANCE - 4*plaintext_integer)

    print("************* Transfer clear ", plaintext_integer, " from my account to Alice *************")
    # Transfer 5 SOD to Alice
    function = contract.functions.contractTransferFromClear(account.address, alice_address.address, plaintext_integer)
    execute_and_check_balance(w3, account, contract, function, user_key, 'contract transfer from clear ', INITIAL_BALANCE - 5*plaintext_integer)

    print("************* Transfer IT ", plaintext_integer, " from my account to Alice *************")
    # Transfer 5 SOD to Alice
    function = contract.functions.transferFrom(account.address, alice_address.address, dummyCT, dummySignature, False) # Dummy function
    func_sig = get_function_signature(function.abi) # Get the function signature
    # Prepare the input text for the function
    ct, signature = prepare_IT(plaintext_integer, user_key, account, contract, func_sig, bytes.fromhex(private_key[2:]))
    # Create the real function using the prepared IT
    function = contract.functions.transferFrom(account.address, alice_address.address, ct, signature, False)
    execute_and_check_balance(w3, account, contract, function, user_key, 'transfer from IT', INITIAL_BALANCE - 6*plaintext_integer)
    
    print("************* Check my allowance *************")
    # Check the remainning allowance
    check_allowance(account, contract, user_key, plaintext_integer*7)

    print("************* Approve IT 50 to my address *************")
    # Approve 50 SOD to this account
    function = contract.functions.approve(account.address, dummyCT, dummySignature) # Dummy function
    func_sig = get_function_signature(function.abi) # Get the function signature
    # Prepare the input text for the function
    ct, signature = prepare_IT(50, user_key, account, contract, func_sig, bytes.fromhex(private_key[2:]))
    function = contract.functions.approve(account.address, ct, signature)
    execute_transaction(w3, account, function)

    print("************* Check my allowance *************")
    # Check that the allowance has changed to 50 SOD
    check_allowance(account, contract, user_key, 50)




if __name__ == "__main__":
    main()

File: examples/scripts/run_erc20.mjs
import fs from 'fs';
import { block_size, decrypt, prepareIT, hexBase } from '../../tools/js/crypto.js';
import solc from 'solc';
import Web3 from 'web3';

const SOLC_VERSION = '0.8.19';
const FILE_PATH = 'ERC20TokenStandardContract.sol';
const PROVIDER_URL = 'http://localhost:7000';
const INITIAL_BALANCE = 500000000

function checkInstalledSolcVersion(solcVersion) {
    const installedVersion = solc.version();
    console.log('Installed Solc version:', installedVersion);
    if (!installedVersion.includes(solcVersion)) {
        throw new Error(`Solc version ${solcVersion} is not installed. Please install it.`);
    } else {
        console.log(`Solc version ${solcVersion} is already installed.`);
    }
}

function readSolidityCode(filePath) {
    if (!fs.existsSync(filePath)) {
        throw new Error(`The file ${filePath} does not exist`);
    }
    return fs.readFileSync(filePath, 'utf8');
}

function compileSolidity(solcVersion, filePath, solidityCode, mpcInstPath, mpcCorePath) {
    const input = {
        language: 'Solidity',
        sources: {
            'MPCInst.sol': { content: fs.readFileSync(mpcInstPath, 'utf8') },
            'MpcCore.sol': { content: fs.readFileSync(mpcCorePath, 'utf8') },
            [filePath]: { content: solidityCode },
        },
        settings: {
            outputSelection: {
                '*': {
                    '*': ['abi', 'metadata', 'evm.bytecode', 'evm.bytecode.sourceMap'],
                },
            },
        },
    };
    const output = JSON.parse(solc.compile(JSON.stringify(input)));
    return output.contracts[filePath][filePath.split('.')[0]];
}

async function deployContract(web3, account, contractABI, contractBytecode, gas_limit=10000000, gas_price='30'){
    const contract = new web3.eth.Contract(contractABI);
    const deployTx = contract.deploy({
        data: contractBytecode,
        arguments: ["Soda", "SOD", INITIAL_BALANCE]
    });
    const deployOptions = {
        data: deployTx.encodeABI(),
        gas: gas_limit,
        from: account.address,
        gasPrice: web3.utils.toWei(gas_price.toString(), 'gwei') // Convert gas price to wei
    };

    try {
        const signedTx = await web3.eth.accounts.signTransaction(deployOptions, account.privateKey);
        const txHash = await new Promise((resolve, reject) => {
            web3.eth.sendSignedTransaction(signedTx.rawTransaction)
                .once('receipt', receipt => resolve(receipt.transactionHash))
                .once('error', reject);
        });
    
        const receipt = await web3.eth.getTransactionReceipt(txHash);
        console.log('Contract deployed at:', receipt.contractAddress);
    
        // Access the deployed contract instance using new keyword
        const deployedContract = new web3.eth.Contract(contractABI, receipt.contractAddress);
        
        // Access the deployed contract's address
        console.log('Deployed contract address:', deployedContract.options.address);
    
        return deployedContract;
        // return new web3.eth.Contract(contractABI, receipt.contractAddress);
    } catch (error) {
        console.error('Error deploying contract:', error);
        throw error;
    }
    
}

async function executeTransaction(web3, account, func, gas_limit=10000000, gas_price='30') {
    try {
        const nonce = await web3.eth.getTransactionCount(account.address);
        const transaction = await func.send({
            from: account.address,
            chainId: 50505050,
            gas: gas_limit,
            gasPrice: web3.utils.toWei(gas_price, 'gwei'),
            nonce: nonce,
        });

        const txnReceipt = await web3.eth.getTransactionReceipt(transaction.transactionHash);
        console.log(`Transaction successful with hash: ${txnReceipt.transactionHash}`);
        return txnReceipt;
    } catch (error) {
        console.error('Error executing transaction:', error);
        throw error;
    }
}

function checkExpectedResult(name, expectedResult, result) {
    if (result === expectedResult) {
        console.log(`Test ${name} succeeded: ${result}`);
    } else {
        throw new Error(`Test ${name} failed. Expected: ${expectedResult}, Actual: ${result}`);
    }
}

function getFunctionSignature(func) {
    const encodedABI = func.encodeABI();
    return Buffer.from(encodedABI.substr(2, 8), 'hex');
}

function decryptValue(myCTBalance, userKey) {

    // Convert CT to bytes
    let ctString = myCTBalance.toString(hexBase);
    let ctArray = Buffer.from(ctString, 'hex');
    while (ctArray.length < 32) { // When the first bits are 0, bigint bit size is less than 32 and need to re-add the bits
        ctString = "0" + ctString;
        ctArray = Buffer.from(ctString, 'hex');
    }
    // Split CT into two 128-bit arrays r and cipher
    const cipher = ctArray.subarray(0, block_size);
    const r = ctArray.subarray(block_size);

    // Decrypt the cipher
    const decryptedMessage = decrypt(userKey, r, cipher);

    // console.log the decrypted cipher
    const decryptedBalance = parseInt(decryptedMessage.toString('hex'), block_size);

    return decryptedBalance;
}

async function executeAndCheckBalance(web3, account, func, contract, user_key, expectedBalance){
    await executeTransaction(web3, account, func);
    // Get my encrypted balance, decrypt it and check if it is equal to the expected balance
    const my_CTBalance = await contract.methods.balanceOf().call({'from': account.address});
    const my_balance = decryptValue(my_CTBalance, user_key);
    checkExpectedResult('transfer', expectedBalance, my_balance);
}

async function checkAllowance(account, contract, user_key, expectedAllowance){
    // Get my encrypted allowance, decrypt it and check if it is equal to the expected allowance
    let allowanceCT = await contract.methods.allowance(account.address, account.address).call({'from': account.address})
    let allowance = decryptValue(allowanceCT, user_key)
    checkExpectedResult('allowance', expectedAllowance, allowance)
}

async function main() {
    checkInstalledSolcVersion(SOLC_VERSION);
    const solidityCode = readSolidityCode("../contracts/" + FILE_PATH);
    
    const mpcInstPath = '../../lib/solidity/MPCInst.sol';
    const mpcCorePath = '../../lib/solidity/MpcCore.sol';
    const compiledContract = compileSolidity(SOLC_VERSION, FILE_PATH, solidityCode, mpcInstPath, mpcCorePath);

    const web3 = new Web3(new Web3.providers.HttpProvider(PROVIDER_URL));
    if (!web3.eth.net.isListening()) {
        console.log("Failed to connect to the Ethereum node.");
        return;
    }
    const SIGNING_KEY = process.env.SIGNING_KEY;
    web3.eth.accounts.wallet.add(SIGNING_KEY);
    const account = web3.eth.accounts.wallet[0];

    let contract = web3.eth.Contract 
    contract = await deployContract(web3, account, compiledContract.abi, compiledContract.evm.bytecode.object);
    
    console.log("************* View functions *************");
    const contractName = await contract.methods.name().call({'from': account.address});
    console.log("Function call result name:", contractName);

    const symbol = await contract.methods.symbol().call({'from': account.address});
    console.log("Function call result symbol:", symbol);

    const decimals = await contract.methods.decimals().call({'from': account.address});
    console.log("Function call result decimals:", decimals);

    const totalSupply = await contract.methods.totalSupply().call({'from': account.address});
    console.log("Function call result totalSupply:", totalSupply);

    const user_key_hex = process.env.USER_KEY;
    const user_key = Buffer.from(user_key_hex, 'hex');

    console.log("************* Initial balance = ", INITIAL_BALANCE, " *************");
    let my_CTBalance = await contract.methods.balanceOf().call({'from': account.address});
    let my_balance = decryptValue(my_CTBalance, user_key);
    checkExpectedResult('balanceOf', INITIAL_BALANCE, my_balance);

    // Generate a new account for Alice
    const alice_address = web3.eth.accounts.create();

    const plaintext_integer = 5;


    console.log("************* Transfer clear ", plaintext_integer, " to Alice *************");
    // Transfer 5 SOD to Alice
    let func = contract.methods.transfer(alice_address.address, plaintext_integer, true);
    await executeAndCheckBalance(web3, account, func, contract, user_key, INITIAL_BALANCE - plaintext_integer)

    console.log("************* Transfer clear ", plaintext_integer, " to Alice *************");
    // Transfer 5 SOD to Alice
    func = contract.methods.contractTransferClear(alice_address.address, plaintext_integer);
    await executeAndCheckBalance(web3, account, func, contract, user_key, INITIAL_BALANCE - 2*plaintext_integer)

    console.log("************* Transfer IT ", plaintext_integer, " to Alice *************")
    // In order to generate the input text, we need to use some data of the function. 
    // For example, the address of the user, the address of the contract and also the function signature.
    // In order to get the function signature as simple as possible, we decided to use a dummy function with dummy inputs and use it to generate the signature.
    // After the signature is generated, we call prepare input text function and get the input text to use in the real function.
    const dummyCT = 0;
    const dummySignature = Buffer.alloc(65);
    func = contract.methods.transfer(alice_address.address, dummyCT, dummySignature, false); // Create dummy function to get the signature for prepare input text
    let hashFuncSig = getFunctionSignature(func);
    // After the function signature is obtained, prepare the input test for the function
    let {ctInt, signature} = prepareIT(plaintext_integer, user_key, account.address, contract.options.address, hashFuncSig, Buffer.from(SIGNING_KEY.slice(2), 'hex'));
    // Create the real function using the prepared IT
    func = contract.methods.transfer(alice_address.address, ctInt, signature, false);
    // Transfer 5 SOD to Alice
    await executeAndCheckBalance(web3, account, func, contract, user_key, INITIAL_BALANCE - 3*plaintext_integer)

    console.log("************* Transfer clear ", plaintext_integer, " from my account to Alice without allowance *************")
    // Trying to transfer 5 SOD to Alice. There is no allowance, transfer should fail
    func = contract.methods.transferFrom(account.address, alice_address.address, plaintext_integer, true)
    // There is no allowance, balance should remain the same
    await executeAndCheckBalance(web3, account, func, contract, user_key, INITIAL_BALANCE - 3*plaintext_integer)

    console.log("************* Approve ", plaintext_integer*10, " to my address *************")
    // Set allowance for this account
    func = contract.methods.approveClear(account.address, plaintext_integer*10)
    await executeTransaction(web3, account, func)
    // Get my allowance to check that the allowance is correct
    checkAllowance(account, contract, user_key, plaintext_integer*10)

    console.log("************* Transfer clear ", plaintext_integer, " from my account to Alice *************")
    // Transfer 5 SOD to Alice
    func = contract.methods.transferFrom(account.address, alice_address.address, plaintext_integer, true)
    await executeAndCheckBalance(web3, account, func, contract, user_key, INITIAL_BALANCE - 4*plaintext_integer)

    console.log("************* Transfer clear ", plaintext_integer, " from my account to Alice *************")
    // Transfer 5 SOD to Alice
    func = contract.methods.contractTransferFromClear(account.address, alice_address.address, plaintext_integer)
    await executeAndCheckBalance(web3, account, func, contract, user_key, INITIAL_BALANCE - 5*plaintext_integer)

    console.log("************* Transfer IT ", plaintext_integer, " from my account to Alice *************");
    // Transfer 5 SOD to Alice
    func = contract.methods.transferFrom(account.address, alice_address.address, dummyCT, dummySignature, false);
    hashFuncSig = getFunctionSignature(func);
    ({ctInt, signature} = prepareIT(plaintext_integer, user_key, account.address, contract.options.address, hashFuncSig, Buffer.from(SIGNING_KEY.slice(2), 'hex')));
    // Create the real function using the prepared IT output
    func = contract.methods.transferFrom(account.address, alice_address.address, ctInt, signature, false);
    await executeAndCheckBalance(web3, account, func, contract, user_key, INITIAL_BALANCE - 6*plaintext_integer);
    
    console.log("************* Check my allowance *************")
    // Check the remainning allowance
    checkAllowance(account, contract, user_key, plaintext_integer*7)

    console.log("************* Approve IT 50 to my address *************")
    // Approve 50 SOD to this account
    func = contract.methods.approve(account.address, dummyCT, dummySignature); // Dummy function to get the signature
    hashFuncSig = getFunctionSignature(func);
    ({ctInt, signature} = prepareIT(50, user_key, account.address, contract.options.address, hashFuncSig, Buffer.from(SIGNING_KEY.slice(2), 'hex')));
    func = contract.methods.approve(account.address, ctInt, signature);
    await executeTransaction(web3, account, func);

    console.log("************* Check my allowance *************")
    // Check that the allowance has changed to 50 SOD
    checkAllowance(account, contract, user_key, 50);
}

main();
